app/scripts/lib/multichain-api/adapters/caip-permission-adapter-permittedChains.test.ts

```diff
@@ -0,0 +1,314 @@
+import { Caip25CaveatValue } from '../caip25permissions';
+import { validNotifications, validRpcMethods } from '../scope';
+import {
+  addPermittedEthChainId,
+  getPermittedEthChainIds,
+  setPermittedEthChainIds,
+} from './caip-permission-adapter-permittedChains';
+
+describe('CAIP-25 permittedChains adapters', () => {
+  describe('getPermittedEthChainIds', () => {
+    it('returns the unique set of EIP-155 chainIds in hex from the CAIP-25 caveat value', () => {
+      const ethChainIds = getPermittedEthChainIds({
+        requiredScopes: {
+          'eip155:1': {
+            methods: [],
+            notifications: [],
+            accounts: ['eip155:1:0x1', 'eip155:1:0x2'],
+          },
+          'eip155:5': {
+            methods: [],
+            notifications: [],
+            accounts: ['eip155:5:0x2', 'eip155:1:0x3'],
+          },
+          'bip122:000000000019d6689c085ae165831e93': {
+            methods: [],
+            notifications: [],
+            accounts: [
+              'bip122:000000000019d6689c085ae165831e93:128Lkh3S7CkDTBZ8W7BbpsN3YYizJMp8p6',
+            ],
+          },
+        },
+        optionalScopes: {
+          'eip155:1': {
+            methods: [],
+            notifications: [],
+            accounts: ['eip155:1:0x1', 'eip155:1:0x4'],
+          },
+          'eip155:10': {
+            methods: [],
+            notifications: [],
+          },
+          'eip155:100': {
+            methods: [],
+            notifications: [],
+            accounts: ['eip155:100:0x100'],
+          },
+        },
+        isMultichainOrigin: false,
+      });
+
+      expect(ethChainIds).toStrictEqual(['0x1', '0x5', '0xa', '0x64']);
+    });
+  });
+
+  describe('addPermittedEthChainId', () => {
+    it('adds an optional scope for the chainId if it does not already exist in required or optional scopes', () => {
+      const result = addPermittedEthChainId(
+        {
+          requiredScopes: {
+            'eip155:1': {
+              methods: [],
+              notifications: [],
+              accounts: ['eip155:1:0x1', 'eip155:1:0x2'],
+            },
+          },
+          optionalScopes: {
+            'eip155:100': {
+              methods: [],
+              notifications: [],
+              accounts: ['eip155:100:0x100'],
+            },
+          },
+          isMultichainOrigin: false,
+        },
+        '0x65',
+      );
+
+      expect(result).toStrictEqual({
+        requiredScopes: {
+          'eip155:1': {
+            methods: [],
+            notifications: [],
+            accounts: ['eip155:1:0x1', 'eip155:1:0x2'],
+          },
+        },
+        optionalScopes: {
+          'eip155:100': {
+            methods: [],
+            notifications: [],
+            accounts: ['eip155:100:0x100'],
+          },
+          'eip155:101': {
+            methods: validRpcMethods,
+            notifications: validNotifications,
+            accounts: [],
```

Comment:
Should this method also add all currently permitted accounts to this scope as well?

Reply:
Shouldn't be necessary. See https://github.com/MetaMask/metamask-extension/pull/26483#discussion_r1739493377